src/Inventory.zig

```diff
@@ -1620,17 +1622,56 @@ pub const Command = struct { // MARK: Command
 			if(side == .server and gamemode != .creative and self.oldBlock.typ != self.newBlock.typ and shouldDropSourceBlockOnSuccess) {
 				for(self.oldBlock.blockDrops()) |drop| {
 					if(drop.chance == 1 or main.random.nextFloat(&main.seed) < drop.chance) {
-						blockDrop(self.pos, drop);
+						blockDrop(self.pos, self.dropDirection, drop);
 					}
 				}
 			}
 		}
 
-		fn blockDrop(pos: Vec3i, drop: main.blocks.BlockDrop) void {
-			for(drop.items) |itemStack| {
-				const dropPos = @as(Vec3d, @floatFromInt(pos)) + @as(Vec3d, @splat(0.5)) + main.random.nextDoubleVectorSigned(3, &main.seed)*@as(Vec3d, @splat(0.5 - main.itemdrop.ItemDropManager.radius));
-				const dir = vec.normalize(main.random.nextFloatVectorSigned(3, &main.seed));
-				main.server.world.?.drop(itemStack.clone(), dropPos, dir, main.random.nextFloat(&main.seed)*1.5);
+		fn blockDrop(_pos: Vec3i, _dir_hint: Vec3d, drop: main.blocks.BlockDrop) void {
+			const currentBlock: Block = main.server.world.?.getBlock(_pos[0], _pos[1], _pos[2]) orelse Block{.typ = 0, .data = 0};
+			var dropVelocity = 4.0 + main.random.nextFloatSigned(&main.seed)*0.5;
+
+			if(!currentBlock.solid() or !currentBlock.collide()) {
+				const dropDir = vec.normalize(Vec3d{
+					main.random.nextDoubleSigned(&main.seed),
+					main.random.nextDoubleSigned(&main.seed),
+					2.0,
+				});
+
+				const sign = std.math.sign(dropDir);
+				const centerOffset = Vec3d{0.5, 0.5, 0.0};
+				const itemOffset = sign*@as(Vec3d, @splat(main.itemdrop.ItemDropManager.radius));
+				const dropPos = @as(Vec3d, @floatFromInt(_pos)) + centerOffset + itemOffset*@as(Vec3d, @splat(2));
+
+				for(drop.items) |itemStack| {
+					main.server.world.?.drop(itemStack.clone(), dropPos, @floatCast(dropDir), dropVelocity);
+				}
+			} else {
+				const sign = std.math.sign(_dir_hint);
+				const factors = Vec3d{
+					@sin(@abs(std.math.atan(_dir_hint[0]))),
+					@sin(@abs(std.math.atan(_dir_hint[1]))),
+					@sin(@abs(std.math.atan(_dir_hint[2]))),
+				};
+				const edgeOffset = _dir_hint + factors*@as(Vec3d, @splat(std.math.sqrt(2) - 1));
```

Comment:
This math seems needlessly complex.
Couldn't you just also send a position hint/box (which would be useful for e.g. torches anyways) to avoid having to retrieve it here from just the direction?